proposals/0000-severe-warnings.rst

```diff
@@ -0,0 +1,157 @@
+-Wsevere – erroring warnings
+============================
+
+.. author:: Joachim Breitner
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/571>`_.
+.. contents::
+
+It seems helpful for some warnings (e.g. ``missing-methods``) to be
+*errors* by default (as if ``-Werror=missing-methods`` was given), but still
+allow them to be turned into warnings (``-Wwarn=missing=methods``) or being disabled
+(``-Wno-missing-methods``). This proposal
+adds a new warning group ``-Wsevere`` (like ``-Wextra`` or ``-Wdefault``) for such warnings
+and upgrades ``missing-methods``,  ``missing-fields`` from ``-Wdefault`` to ``-Wsevere``
+(fixing `#544 <https://github.com/ghc-proposals/ghc-proposals/issues/544>`_).
+
+Motivation
+----------
+
+* We can start for the motivation for making ``missing-methods`` an error by default, copying from @phadej’s
+  `#544 <https://github.com/ghc-proposals/ghc-proposals/issues/544>`_:
+  
+    Currently omitting method implementations in type-class instances
+    is just a warning, but it is enabled by default.
+    Therefore one could argue that adding a new method to existing
+    type-class is not a breaking change.
+    However, it very much is, but
+
+    *  ecosystem is not forced to catch up in any timely matter
+    *  the breakage is hard to assess
+
+  and (to me even more compelling):
+  
+    FWIW, base-4.18.0.0 is going to have
+
+    Add default implementation of `(<>)` in terms of `sconcat` and `mempty` in terms of `mconcat`.
+    which makes it possible to write::
+
+      instance Semigroup Foo
+      instance Monoid Foo
+
+    and that instance's mempty will loop.
+    Currently GHC will just warn if someone writes such (new) code by accident.
```

Comment:
I'm confused by this, if the methods have default implementations, will GHC really warn?

Reply:
Good point. Maybe a MINIMAL pragma is involved (but they don't span multiple classes, do they). @phadej?

Reply:
with MINIMAL pragma GHC warns. That's the point of MINIMAL pragma, to allow circular default definitions but make compiler still warn.

Reply:
Ah, right, so the circular definitions don't actually span multiple classes. Makes sense!